### ISSUE TITLE
Different data type after 'CASE WHEN xxx THEN' and 'ELSE'

### ISSUE THREAD

**Issue by cornmonster** [tableau/connector-plugin-sdk#191]
**About You:**
Name: Jon
Company: MaxCompute

**Your question:**
Hi,

The test case 'logical.case.null' generates a SQL statements as follows:
```
SELECT (CASE WHEN ((8 + DATEDIFF(`calcs`.`date1`,NEXT_DAY(CAST(`calcs`.`date1` AS DATE),'SU'))) IN (7, 1)) THEN CAST(NULL AS TIMESTAMP) ELSE `calcs`.`date1` END) AS `temp_test__4257957843__0_`
FROM `odpsdemo_dev`.`calcs` `calcs`
GROUP BY (CASE WHEN ((8 + DATEDIFF(`calcs`.`date1`,NEXT_DAY(CAST(`calcs`.`date1` AS DATE),'SU'))) IN (7, 1)) THEN CAST(NULL AS TIMESTAMP) ELSE `calcs`.`date1` END)
```

The data type of the field after 'CASE WHEN xxx THEN' is TIMESTAMP while the data type of the field `calcs`.`date1` after 'ELSE' is DATE.

We have tried to run the same query on hive and hive could not accept this query as well. 

So, is it possible for us to safely skip this test case? Or if this test case is really important, what should we do to get it passed?

Thank you very much for your time!
Jon

**Comment by liyunbao** [comment]
Hi Jon,

The root cause of this failure is similar to https://github.com/tableau/connector-plugin-sdk/issues/189. The generated sql is based on Hive12Dialect but it seems like the query is not working properly on your hive.

One possible solution is:
Can you run the sql below on your hive and see if it works. This query is functional equivalent to the one your pasted, the difference it is generated based on HiveDialect, not Hive12Dialect.  If it works, can you update your base class in you dialect tdd file as "HiveDialect" and see if it can pass the test.
'''
    SELECT (CASE WHEN ((1 + IF(7<0, PMOD((IF(7<0, PMOD(DATEDIFF(TO_DATE(`calcs`.`date1`), TO_DATE('1995-01-01')),7)-7, PMOD(DATEDIFF(TO_DATE(`calcs`.`date1`), TO_DATE('1995-01-01')), 7)) + 7),7)-7, PMOD((IF(7<0, PMOD(DATEDIFF(TO_DATE(`calcs`.`date1`), TO_DATE('1995-01-01')),7)-7, PMOD(DATEDIFF(TO_DATE(`calcs`.`date1`), TO_DATE('1995-01-01')), 7)) + 7), 7))) IN (7, 1)) THEN CAST(NULL AS TIMESTAMP) ELSE `calcs`.`date1` END) AS `temp_test__4257957843__0_`
      FROM `odpsdemo_dev`.`calcs` `calcs`
      GROUP BY 1
'''

Thanks,
Liyun

**Comment by liyunbao** [comment]
This issue is fixed and the fix will be available on Tableau 2019.4 release. Hence close it.
